### PR TITLE
Implement Analytical Derivatives on `CompoundModel`

### LINKED ISSUE #3783: Implement deriv and fit_deriv for compound models
Compound models currently do not implement analytic derivatives (with respect to their parameters) through a `fit_deriv` method.  This hasn't been a showstopper, since all of the optimizers that use the derivative can use an estimate if an analytic expression is not provided.

However, this is much less straightforward than mere evaluation of compound models.  It will have compute partial derivatives with respect to each parameter in the compound model, and will have to take into account which operators each model is combined with, and implement the appropriate product rule, quotient rule, chain rule, etc.  There's nothing especially complicated about the math here, but getting the implementation right may be a bit tedious.  The most annoying part about this is that we'll have to implement derivatives with respect to the inputs for each model (`fit_deriv` implements partial derivatives w.r.t. the fittable parameters).  Perhaps there should be a separate issue for this.

To provide some hints for anyone who might like to tackle this, model _evaluation_ is currently implemented by passing a dictionary mapping operators (given as strings) to functions that implement that operator between two models:

https://github.com/astropy/astropy/blob/cee42f3923a4fa9ea4d20c206baa9a9789e9defd/astropy/modeling/core.py#L1379

For example, the `'+'` operator is simply implemented as "A(x) + B(x)" where A and B are the `evaluate` methods of two models (there is a little more complexity involved to support things like multiple inputs/outputs.  You can see how this is implemented in `_make_arithmetic_operator` above.  

Similarly the `'|'` operator is implemented to evaluate "B(A(x))" (again modulo some slight complications for generalization, and parameter handling--let me know if these need better explanation).

So for evaluating `fit_deriv`, if you have a model like "A + B", the `'+'` operation would have to be implemented like `A.fit_deriv(x) + B.fit_deriv(x)`.  That's easy enough.  But for implementing `'|'` we'll need a derivative method.  So "A | B" could be implemented like "B.deriv(A(x)) \* A.fit_deriv(x)".  Though things get more complicated for models of multiple arguments, etc.  Not also that in this example, if "B" is itself a compound model it will have to use a `B.deriv` for the entire compound model; i.e. this would have to work recursively.

### PR DESCRIPTION
**Pull request by Cadair**
### Description

This pull request adds support for computing analytical derivatives on `CompoundModel` for the +, -, * and / operators.
Others are possible, but haven't been attempted yet.

The main motivation for this PR is to *substantially* improve the performance of fitting these models. The [profiling script](https://github.com/aperiosoftware/SPICE-Model-Fitting-Reports) we are using which fits a `Const + Gaussian1D` model in parallel went from 28.1s to 19.8s so a 1.4x speedup.

Probably Fixes #3783, but that might have a slightly larger scope

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/modeling/tests/test_compound.py [hunk 1]
```diff
@@ -1216,3 +1216,55 @@ def test_fit_mixed_recursive_compound_model_with_mixed_units():
...
+    return (up - down) / (2 * delta)
+
+
+@pytest.mark.parametrize(
+    "model",
+    [
+        Gaussian1D(5, 2, 3) + Linear1D(2, 3),
+        Gaussian1D(5, 2, 3) - Linear1D(2, 3),
+        Polynomial1D(2) * Gaussian1D(),
+        Polynomial1D(2) / Gaussian1D(),
```
Comment (Cadair) [inline astropy/modeling/tests/test_compound.py:1255]: This example seemed to fail in a notebook yesterday:
```suggestion
        Polynomial1D(2) / Gaussian1D(),
        Polynomial1D(degree=0, c0=664.3349) + Gaussian1D(amplitude=2000, mean=195),
```

**Comment by astrofrog** [comment]
The Polynomial2D thing was a red herring, fitters will just call ``fit_deriv`` with 1D arrays even when there are multiple dimensions, so this won't matter. We are down to three test failures: one because there is a difference of NaN vs Inf for the numerical vs analytical derivatives, and the other one is a slight numerical difference in the final model, which might just be a numerical issue, so might be safe to increase the relative tolerance?

**Review by astrofrog — APPROVED** [review]
This looks good to me! Presumably @nden or @perrygreenfield should review too.

**Review by WilliamJamieson — APPROVED** [review]
These look good to me. 

One comment, we should look into supporting the computation through function composition. This should be doable as that is just chain rule.

**Comment by neutrinoceros** [comment]
Looks to me that this can now go in. I've opened #17250 as a reminder of @WilliamJamieson's suggestion to expand the scope later on.
If no one complains about it I intend to merge this later today.

**Comment by Cadair** [comment]
@neutrinoceros :shipit::question:

**Comment by neutrinoceros** [comment]
Screenshot 2024-10-25 at 20 16 22

**Comment by neutrinoceros** [comment]
This PR caused a regression in 7.0.0: https://github.com/astropy/astropy/issues/17509
Lessons learned:
- I should remember to squash-merge more often
- I should refrain from merging sizeable features literal hours ahead of feature freeze 🥶

**Comment by Cadair** [comment]
> I should refrain from merging sizeable features literal hours ahead of feature freeze 🥶

Better than hours afterwards ;)

**Pull request merged**